types/redux-first-router/index.d.ts

```diff
@@ -12,34 +12,34 @@ import {
     Store,
     Reducer,
     Middleware,
-    GenericStoreEnhancer
+    StoreEnhancer
 } from 'redux';
 import { History } from 'history';
 
 export type Nullable<T> = T | null | undefined;
 
-export type StateGetter = () => object;
+export type StateGetter<TState = any> = () => TState;
 
 export type RouteString = string;
 
-export type RouteThunk = (
-    dispatch: Dispatch<any>,
-    getState: StateGetter
+export type RouteThunk<TState = any> = (
+    dispatch: Dispatch<TState>,
```

Comment:
Dispatch takes an action as input. Signature could be completely different.
https://github.com/reactjs/redux/blob/master/index.d.ts#L103

Reply:
In `redux@4.0`, the typing is 

```typescript
export interface Dispatch<D = Action> {
    <A extends D>(action: A): A;
}
```
([source](https://github.com/reactjs/redux/blob/f353e613e02a6a66e5b4944bb849e80c141b0f8a/index.d.ts#L103))

yes, but in `redux@3.7.2` it is defined as;

```typescript
export interface Dispatch<S> {
    <A extends Action>(action: A): A;
}
```
([source](https://github.com/reactjs/redux/blob/8f60ba321e8ba5fa71d60fa35573c2cdf9c0d852/index.d.ts#L107))

where `S` is `State` (see def of [`MiddlewareAPI`](https://github.com/reactjs/redux/blob/8f60ba321e8ba5fa71d60fa35573c2cdf9c0d852/index.d.ts#L267)), which I'd agree does not make a lot of sense. 

So with `3.7` definitions, it'd be perfectly fine to define the `dispatch`  in `RouteThunk` as `Dispatch<any>` since the type parameter doesn't affect the function at all. 

With the `4.0` typings i take it the idea is to use the dispatch typing as `Dispatch<MyAction | ThunkAction | MyOtherAction>` to basically typecheck what you call `dispatch()`with? In that case it would make sense to have `RouteThunk` take a Action as an optional type parameter. However more changes are needed to the definitions to adjust for redux 4.0.

How do you think we should do with `RouteThunk` for now?

Reply:
I would not change `RouteThunk`. The amount of types could be a lot and you could widen the type yourself inside the declaration itself.

You could refer to the Action type inside Redux.
https://github.com/reactjs/redux/blob/master/index.d.ts#L19

`dispatch: Dispatch<Action>`